cronjobs/src/commands/_git_export_git_tools.py

```diff
@@ -0,0 +1,194 @@
+            print(f"Resetting local branch {branch_name} to remote {remote_ref_name}")
+            local_branch.set_target(remote_branch.target)
+
+    # Delete local tags that are not on remote
+    origin = repo.remotes[REMOTE_NAME]
+    remote_tags = {
+        obj["name"]
+        for obj in origin.ls_remotes(callbacks=callbacks)
+        if obj["name"].startswith("refs/tags/") and not obj["local"]
+    }
+    for ref in repo.references:
+        if not ref.startswith("refs/tags/") or ref in remote_tags:
+            continue
+        print(f"Delete local tag {ref}")
+        repo.references.delete(ref)
+
+
+def push_mirror(
+    repo: pygit2.Repository,
+    branches: list[str],
+    tags: list[str],
+    callbacks: pygit2.RemoteCallbacks,
+):
+    """
+    An equivalent of `git push --force --mirror` for branches + tags only.
+    """
+    to_push = []
+    # 1) Force update all local branches
+    for b in sorted(branches):
+        to_push.append(f"+{b}:{b}")
+    # 2) Force update all local tags
+    for t in sorted(tags):
+        to_push.append(f"+refs/tags/{t}:refs/tags/{t}")
+
+    if not to_push:
+        print("Everything up-to-date.")
+        return
+
+    remote = repo.remotes[REMOTE_NAME]
+    print(f"Pushing to remote {remote.url}:\n - {'\n - '.join(to_push)}")
+    # This is the critical bit: non-fast-forward updates require the '+' force.
+    # The deletions use the ':refs/...'; '+' is ignored for deleted refspecs.
+    remote.push(to_push, callbacks=callbacks)
+
+
+def make_lfs_pointer(sha256_hex: str, size: int) -> pygit2.Oid:
+    """
+    Create a Git LFS pointer blob with the given object id and size.
+    """
+    pointer = (
+        "version https://git-lfs.github.com/spec/v1\n"
+        f"oid sha256:{sha256_hex}\n"
+        f"size {size}\n"
+    )
+    return pointer.encode("ascii")
+
+
+def parse_lfs_pointer(data: bytes) -> tuple[str, int]:
+    """
+    Parse a Git LFS pointer blob and return the object id and size.
+    """
+    lines = data.decode("ascii").strip().split("\n")
+    sha256_hex = size = None
+    for line in lines:
+        if line.startswith("oid sha256:"):
+            sha256_hex = line.split(":", 1)[1].strip()
+        elif line.startswith("size "):
+            size = int(line.split(" ", 1)[1].strip())
+    if sha256_hex is None or size is None:
+        raise ValueError(f"Invalid LFS pointer: {data}")
+    return sha256_hex, size
+
+
+def iter_tree(
+    repo: pygit2.Repository, tree: pygit2.Tree, prefix=""
+) -> Generator[tuple[str, pygit2.Oid], None, None]:
+    """
+    Iterate over the entries in a Git tree, and return their paths and IDs.
+    """
+    for entry in tree:
+        path = f"{prefix}{entry.name}"
+        if entry.type == pygit2.GIT_OBJECT_BLOB:
+            yield path, entry.id  # file
+        elif entry.type == pygit2.GIT_OBJECT_TREE:  # descend
+            yield from iter_tree(repo, repo[entry.id], prefix=path + "/")
+
+
+def tree_upsert_blobs(
```

Comment:
From my experimentation with the `git2` crate, I'd have expected we use an in-memory index and record all files in that index before creating all tree objects in a single operation by writing the index to the object database. But, alas, in-memory indexes don't seem to be supported in pygit2.

Reply:
Feel free to ignore this comment. I'm just complaining that pygit2 doesn't offer all the nice things I saw in the Rust git2 crate. This whole function would be far simpler using an in-memory index, since then the library does most of the work for us. We'd just add the files one by one in any order using  `git_index_add_from_buffer()`, and finally we'd call `git_index_write_tree()` to create all the trees in a single operation.